src/api.js

```diff
@@ -17,33 +17,410 @@ const getMessages = args => {
+    while (ids_next) {
+      if (ids_next.length === 1) {
+        const next_message = composition_msgs.splice(
+          findTargetIndex(ids_next[0], composition_msgs),
+          1
+        )[0];
+        pipeline.messages.push(next_message);
+        if (next_message.groupId) {
+          pipeline.groupId = next_message.groupId;
+          if (next_message.pipeTarget) {
+            pipeline.pipeTarget = next_message.pipeTarget;
+          }
+          ids_next = null;
+        } else {
+          if (next_message.pipeTarget) {
+            ids_next = next_message.pipeTarget.map(pipe_element => pipe_element.messageId);
+          } else {
+            ids_next = null;
+          }
+        }
+      } else {
+        const group = assembleGroup(ids_next, composition_msgs);
+        pipeline.messages.push(group);
+        if (group.pipeTarget) {
+          ids_next = group.pipeTarget.map(pipe_element => pipe_element.messageId);
+        } else {
+          ids_next = null;
+        }
+      }
+    }
+    return addDetails(pipeline);
+  }
+
+  /**
+   * Returns for each given id the list of group_ids contained in the message with the given id and its pipe_target
+   * @param start_ids
+   * @param composition_msgs
+   * @returns {*[]}
+   */
+  function findMessagesGroupIds(start_ids, composition_msgs) {
+    const composition_group_ids = [];
+    start_ids.forEach(id => {
+      const msg_group_ids = [];
+      let msg = composition_msgs[findTargetIndex(id, composition_msgs)];
+      if (msg.groupId) {
+        msg_group_ids.push(msg.groupId);
+      }
+      while (msg.pipeTarget) {
+        msg = msg.pipeTarget[0];
+        if (msg.groupId) {
+          msg_group_ids.push(msg.groupId);
+        }
+      }
+      composition_group_ids.push(msg_group_ids);
+    });
+    return composition_group_ids;
+  }
+
+  /**
+   * Returns the group_id of the group that starts with the messages with the given ids
+   * @param start_ids
+   * @param composition_msgs
+   * @returns {any}
+   */
+  function findGroupId(start_ids, composition_msgs) {
+    // We find the group's group_id thanks to this property :
+    // The group's groupId is the first group_id that all first messages + their pipe targets have in common
+    const composition_group_ids = findMessagesGroupIds(start_ids, composition_msgs);
+
+    function isGroupId(group_id) {
+      return composition_group_ids.every(msg_group_ids => msg_group_ids.includes(group_id));
+    }
+
+    for (const group_id of composition_group_ids[0]) {
+      if (isGroupId(group_id)) {
+        return group_id;
+      }
+    }
+    throw 'Invalid composition';
+  }
+
+  /**
+   * Assemble the group starting with the given ids
+   * @param start_ids
+   * @param composition_msgs
+   * @returns {*}
+   */
+  function assembleGroup(start_ids, composition_msgs) {
+    const group = { type: 'group', messages: [] };
+    const groupId = findGroupId(start_ids, composition_msgs);
```

Comment:
I don't understand why this is so complicated. Same for the code after. Is it to manage group of groups?